### PR TITLE
[12.0][MIG] - migration l10n_eu_nace

### PR DESCRIPTION
**Pull request by sbejaoui**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please check Travis.

**Comment by pedrobaeza** [comment]
@sbejaoui please start from #39 instead of 10.0 branch, as this PR contains a good refactoring of the module isolating the NACE tags from partner tags.

**Comment by sbejaoui** [comment]
@pedrobaeza , 

Travis fail on base_iso3166 tests. Can you help me fix it?

The #39 remove data files which essential for this module [l10n_fr_naf_ape](https://github.com/OCA/l10n-france/pull/169). As the PR is not finished yet, and does not take compatibility with other modules, I was thinking that I can base on 10.0 and forward-port the improvements later.

**Review by alexis-via — APPROVED** [review]
Just 2 small remarks : replace "OpenERP" by "Odoo" in the rst files and remove the "u" (for unicode) in __manifest__.py now that we use python 3.

**Review by remi-filament — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Is this starting from #39 as commented?

**Comment by legalsylvain** [comment]
@pedrobaeza, your question has been answered 8 monthes ago : #39 removes all the data and it seems to be a problem for @sbejaoui.
https://github.com/OCA/community-data-files/commit/1816eeb5e196760df1d297898a16966aa99a3fe6

What do you think about that topic ?
Cc @alexis-via

**Comment by pedrobaeza** [comment]
Well, I still think this needs that refactoring before getting 12.0 branch, or we won't be able to change the structure in stable.

**Comment by legalsylvain** [comment]
Well, I don't know this module at all, so I'm abstain regarding the code. but What i see is a PR for V12 that seems finished and have 2 approvals, and in the meatime a WIP PR that doesn't change since 8 monthes, and has to be fixed, regarding data.
Not sure that it's a good idea to prevent merging this PR.
In the meantime, other PR are waiting this one, like https://github.com/OCA/l10n-france/pull/169

@sbejaoui : could you check why travis is red ? 

regards.

**Comment by remi-filament** [comment]
@legalsylvain Travis failed because of an issue with base_iso3166, which is apparently solved since PR #31 merge
I restarted the build of Travis and it ended OK that time.

**Comment by sbejaoui** [comment]
Hi guys, 
Can we merge this one please?
Kind regards.

**Comment by pedrobaeza** [comment]
As said, I prefer the switch of approach, having a dedicated model. I hadn't time for doing it myself and change l10n_fr_ape for not having excuses for not doing it and I plan to do it next week, but if you want to do it, it's welcome.

**Comment by sbejaoui** [comment]
> As said, I prefer the switch of approach, having a dedicated model. I hadn't time for doing it myself and change l10n_fr_ape for not having excuses for not doing it and I plan to do it next week, but if you want to do it, it's welcome.

Hi Pedro,
We have this need for one of our customers and my colleague @qgroulard will start work on it ASAP. I prefer that we don't mix this refactoring in the migrations PR.

In the same context we need to merge [l10n-france\#169](https://github.com/OCA/l10n-france/pull/169).

**Comment by pedrobaeza** [comment]
@sbejaoui but that's the problem: we can't merge the migration, or we will be slaved in this version of that approach / data structure, as the excuse of having other OCA modules depending on previous data structure will raise, so we need to do the refactoring at the same time than the migration.

**Comment by sbejaoui** [comment]
> @sbejaoui but that's the problem: we can't merge the migration, or we will be slaved in this version of that approach / data structure, as the excuse of having other OCA modules depending on previous data structure will raise, so we need to do the refactoring at the same time than the migration.

make sense, we will do that.

**Comment by sbejaoui** [comment]
Hi @pedrobaeza ,
I pushed a commit for the separation. Can you check please.
Thanks.

**Review by njeudy — APPROVED** [review]

**Review by yajo — APPROVED** [review]
@pedrobaeza can you review please?

**Comment by marylla** [comment]
@sbejaoui @pedrobaeza Hi there, the PR seems to be ready for a while now. Can we merge?

File: l10n_eu_nace/migrations/12.0.2.0.0/pre-migration.py
Comment (yajo) [inline l10n_eu_nace/migrations/12.0.2.0.0/pre-migration.py]: Without this patch, possibly this data would be deleted as the module gets updated, and the post migration script would fail:

```suggestion
        SET name=concat('old_', name), noupdate=1
```

File: l10n_eu_nace/migrations/12.0.2.0.0/pre-migration.py
Comment (yajo) [inline l10n_eu_nace/migrations/12.0.2.0.0/pre-migration.py]: Just to be sure:

```suggestion
        WHERE model = 'res.partner.category'
        AND module='l10n_eu_nace'
```

File: l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py
Comment (yajo) [inline l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py]: This is possibly a bottleneck if you have lots of partners. Try this:

```suggestion
# Copyright 2019 ACSONE SA/NV
# Copyright 2020 Tecnativa - Jairo Llopis
# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).

from openupgradelib import openupgrade

from odoo import SUPERUSER_ID, api


@openupgrade.migrate()
def migrate(env):
    nace_categories = env['ir.model_data'].search([
        ('module', '=', 'l10_eu_nace'),
        ('model', '=', 'res.partner.category'),
    ])
    for category_imd in nace_categories:
        partners = env['res.partner'].search([
            ('category_id', 'in', [category_imd.res_id]),
        )
        partners.write({
            'nace_id': env.ref('l10n_eu_nace.%s' % category_imd.name.replace('old_', '', 1)).id,
        })
```

Benefits:

1. Single search for all categories and their XMLID names.
1. Write all partners in a category in batch.
1. More readable.

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]:
```suggestion
    _description = 'European NACE partner category'
```

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: This is used by default to filter:

```suggestion
    active = fields.Boolean(index=True, default=True)
```

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: This is used as default order:

```suggestion
    code = fields.Char(index=True)
```

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: This will be used to search. Also it should be translatable, right?:

```suggestion
    name = fields.Char(index=True, translate=True)
```

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: I'm more in favor of declaring a computed `display_name` field, for readability, but OK. :wink:

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: m2o always should be indexed: https://github.com/odoo/odoo/pull/39332

```suggestion
    parent_id = fields.Many2one(comodel_name="res.partner.nace", index=True)
```

File: l10n_eu_nace/models/res_partner.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner.py]: Always index these: https://github.com/odoo/odoo/pull/39332

```suggestion
    nace_id = fields.Many2one(comodel_name="res.partner.nace", string="NACE", index=True)
```

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: :no_entry_sign: This is wrong, sorry. Try this:

```suggestion
    def name_search(self, name, args=None, operator='ilike', limit=100):
        args = list(args or [])
        args += ['|', ('name', operator, name), ('code', 'ilike', name)]
        return super().name_search(name, args, operator, limit)
```

File: l10n_eu_nace/readme/CONTRIBUTORS.rst
Comment (yajo) [inline l10n_eu_nace/readme/CONTRIBUTORS.rst:4]: I wrote enough code in this review :sweat_smile: 

```suggestion
* Souheil Bejaoui <souheil.bejaoui@acsone.eu>
* Jairo Llopis <jairo.llopis@tecnativa.com>
```

File: l10n_eu_nace/views/res_partner.xml
Comment (yajo) [inline l10n_eu_nace/views/res_partner.xml]: I don't see the need to modify this XMLID:

```suggestion
    <record model="ir.ui.view" id="view_partner_form">
```

**Review by yajo — CHANGES_REQUESTED** [review]
Some blocking points.

@Tecnativa TT17678

**Comment by sbejaoui** [comment]
Hi @Yajo ,

I applied all you suggestions, can you update your review.

Thanks.

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: Oops, I forgot this nasty one:

```suggestion
        args = ['|', ('name', operator, name), ('code', 'ilike', name)] + args
```

**Review by yajo — CHANGES_REQUESTED** [review]
@pedrobaeza Please could you confirm us these migration scripts are gonna work?

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: This is wrong, sorry:

```suggestion
        args = ['|', ('name', operator, name), ('code', 'ilike', name)] + args
```

**Review by yajo — CHANGES_REQUESTED** [review]

**Review by yajo — APPROVED** [review]
There's [a failure in Travis](https://travis-ci.org/github/OCA/community-data-files/jobs/696406353#L1026-L1040), please check it.

**Comment by sbejaoui** [comment]
> There's [a failure in Travis](https://travis-ci.org/github/OCA/community-data-files/jobs/696406353#L1026-L1040), please check it.

Fixed and squashed

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: Oh please, not again! :grin: 

Look, I think you're not understanding my previous comments on this part. You should **always** put trusted parts of the domain **before** untrusted ones. It's a safety measure. Actually this code is buggy and if you look close at it you'll see it.

Check this talk on the matter: https://youtu.be/27K-LmTPnZQ?t=2965

You can also use `odoo.osv.expression.OR/AND` if you don't want to worry about the details.

**Review by yajo — CHANGES_REQUESTED** [review]

File: l10n_eu_nace/models/res_partner_nace.py
Comment (sbejaoui) [reply l10n_eu_nace/models/res_partner_nace.py]: It's now when I get your point :pensive: .

I didn't think at all about that as it's already done in super (args +=) :confused:, 

https://github.com/odoo/odoo/blob/c881e3fe2a8b3db43fa9613a52b996fdfab21392/odoo/models.py#L1663-L1676

Comment (sbejaoui) [reply l10n_eu_nace/models/res_partner_nace.py]: I pushed a new commit (even it has no effect)

File: l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py
Comment (pedrobaeza) [inline l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py]: Shouldn't you look only for records with `old` prefix?

File: l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py
Comment (pedrobaeza) [inline l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py]: Better to do this in SQL and in one shot (more or less this query - not exact):
```
UPDATE res_partner rp
SET nace_id = imd2.res_id
FROM ir_model_data imd1
LEFT JOIN ir_model_data imd2
    ON  imd2.module = imd1.module
    AND imd2.name = substring(imd1.name, 4)
WHERE imd1.name LIKE 'old_%'
    AND rp.category_id = imd1.res_is
```

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: This code is buggy. Please remove it, or explain to me what you're trying to do so I can help you fix it :wink: 

```suggestion
```

**Review by yajo — CHANGES_REQUESTED** [review]

File: l10n_eu_nace/models/res_partner_nace.py
Comment (sbejaoui) [reply l10n_eu_nace/models/res_partner_nace.py]: Please check to understand it, https://github.com/odoo/odoo/blob/c881e3fe2a8b3db43fa9613a52b996fdfab21392/odoo/models.py#L1663-L1676

in super, odoo add to args `[(self._rec_name, operator, name)]`. The OR here is needed to end with `['|', ('code', operator, name), (self._rec_name, operator, name)]`

you removed it in your last commit and break the name search.

Comment (yajo) [reply l10n_eu_nace/models/res_partner_nace.py]: But that's an unsafe way to do it, because if you get some `args`, the `|` part will only work for the 1st 2 leafs.

If you really want to avoid that part, what you have to do is ... (see below)

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: ... this:

```suggestion
        return super().name_search(name or self._rec_name, domain, operator, limit)
```

And define a `_rec_name` in the model (there's none AFAICS).

**Comment by rafaelbn** [comment]
Please @sbejaoui , could we finish this? We are in a hurry

File: l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py
Comment (sbejaoui) [reply l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py]: I did it as you suggested but couldn't test it as all our db are migrated. It will be nice if you can run it on existing data.

File: l10n_eu_nace/models/res_partner_nace.py
Comment (sbejaoui) [reply l10n_eu_nace/models/res_partner_nace.py]: I don't get you, 

argument name contain the search value, in case it's empty it will search for nace with name "display_name"?

IMO, there is no way to override name_search without changing the domain in unsafe way due to _name_search implementation in base model.

Avoiding the security leak has no added value as the leak is in base model

This explain why many of OCA and Odoo modules when overriding name_search don't send back to super and return the result of normal search and for the same reasons you last fix suggestions broke the tests.

So, I decide to give up on name_search and use a computed with a dedicated search method

**Comment by sbejaoui** [comment]
> Please @sbejaoui , could we finish this? We are in a hurry

sorry for the late, I believe it's all good now

File: l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py
Comment (yajo) [inline l10n_eu_nace/migrations/12.0.2.0.0/post-migration.py]: Oops, a little big bug here:

```suggestion
                AND imd2.name = substring(imd1.name, 5)
```

FTR:

```sql
devel=# select substring('old_hi', 4);
 substring 
-----------
 _hi
(1 row)

devel=# select substring('old_hi', 5);
 substring 
-----------
 hi
(1 row)
```

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: Oops there's something weird.

```suggestion
        return domain
```

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: Translations are applied later by Odoo.

```suggestion
        ("ref_code", "unique (code)", "NACE Code must be unique!")
```

File: l10n_eu_nace/models/res_partner_nace.py
Comment (yajo) [inline l10n_eu_nace/models/res_partner_nace.py]: Don't you think this should be the opposite?

I mean: the short method is quick, while the long method is possibly a bottleneck (recurses over parents and then reverses a list for each record, both expensive operations).

So, I think that the short method should be used unless explicitly asked by `nace_display=long` context. This way we don't penalize performance unless needed.

What do you think?

BTW, this solution of a computed searchable field is way better than the previous one, good idea!

**Review by yajo — CHANGES_REQUESTED** [review]
This looks much better, but the new code has its new bugs.

Be patient, we're getting close! Thanks.

File: l10n_eu_nace/models/res_partner_nace.py
Comment (sbejaoui) [reply l10n_eu_nace/models/res_partner_nace.py]: I agree with you that's better to use `nace_display=long` but it can break inheritance if it's used in some custom modules as this option was there since previous versions.

Comment (yajo) [reply l10n_eu_nace/models/res_partner_nace.py]: It's not a problem. We are migrating the module, it's the perfect timing to do breaking changes. :wink:

Comment (sbejaoui) [reply l10n_eu_nace/models/res_partner_nace.py]: done

**Review by yajo — APPROVED** [review]

File: l10n_eu_nace/i18n/ru.po
Comment (yajo) [inline l10n_eu_nace/i18n/ru.po]: This file is broken: https://travis-ci.org/github/OCA/community-data-files/jobs/701976078#L1278-L3835

Please delete it for the bot to be able to continue with the merge.

File: l10n_eu_nace/i18n/hr.po
Comment (yajo) [inline l10n_eu_nace/i18n/hr.po]: This file is broken: https://travis-ci.org/github/OCA/community-data-files/jobs/701976078#L3837-L3866

Please delete it so the bot can continue with the merge.

**Review by yajo — CHANGES_REQUESTED** [review]

File: l10n_eu_nace/i18n/hr.po
Comment (sbejaoui) [reply l10n_eu_nace/i18n/hr.po]: done

File: l10n_eu_nace/data/make_data.py
Comment (yajo) [inline l10n_eu_nace/data/make_data.py:95]: FTR those failing translations are probably caused by this line, which should escape properly special characters such as `\n` or `"`.

**Review by yajo — APPROVED** [review]

**Pull request merged**
